mailbox/api/src/test/java/org/apache/james/mailbox/events/EventDeadLettersContract.java

```diff
@@ -0,0 +1,405 @@
+                .isEqualTo(EVENT_ID_1);
+        }
+
+        @Test
+        default void storeShouldIgnoreStoreDuplicatedEventsPerGroup() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+
+            assertThat(eventDeadLetters.failedEventIds(GROUP_A).toStream())
+                .containsExactly(EVENT_ID_1);
+        }
+
+        @Test
+        default void storeShouldKeepConsistencyWhenConcurrentStore() throws Exception {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            ConcurrentHashMap<Integer, Group> groups = concurrentGroups();
+            Multimap<Integer, Event.EventId> storedEventIds = Multimaps.synchronizedSetMultimap(HashMultimap.create());
+            runConcurrent((threadNumber, step) -> {
+                Event.EventId eventId = Event.EventId.random();
+                storedEventIds.put(threadNumber, eventId);
+                eventDeadLetters.store(groups.get(threadNumber), event(eventId));
+            });
+
+            groups.forEach((groupId, group) -> {
+                Group storedGroup = groups.get(groupId);
+                assertThat(eventDeadLetters.failedEventIds(storedGroup).toStream())
+                    .hasSameElementsAs(storedEventIds.get(groupId));
+            });
+        }
+    }
+
+    interface RemoveContract extends EventDeadLettersContract {
+
+        @Test
+        default void removeShouldThrowWhenGroupIsNull() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            assertThatThrownBy(() -> eventDeadLetters.remove(NULL_GROUP, EVENT_ID_1))
+                .isInstanceOf(IllegalArgumentException.class);
+        }
+
+        @Test
+        default void removeShouldThrowWhenEventIdIsNull() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            assertThatThrownBy(() -> eventDeadLetters.remove(GROUP_A, NULL_EVENT_ID))
+                .isInstanceOf(IllegalArgumentException.class);
+        }
+
+        @Test
+        default void removeShouldThrowWhenBothGroupAndEventIdAreNull() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            assertThatThrownBy(() -> eventDeadLetters.remove(NULL_GROUP, NULL_EVENT_ID))
+                .isInstanceOf(IllegalArgumentException.class);
+        }
+
+        @Test
+        default void removeShouldRemoveMatched() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+            eventDeadLetters.store(GROUP_A, EVENT_2).block();
+
+            eventDeadLetters.remove(GROUP_A, EVENT_1.getEventId()).block();
+
+            assertThat(eventDeadLetters.failedEventId(GROUP_A, EVENT_1.getEventId()).block())
+                .isNull();
```

Comment:
Our nice new API is returning null????

Can we seriously consider returning an Optional here?

Reply:
blockOptional ?

Reply:
there is no blockOptional() in Mono